openhcl/virt_mshv_vtl/src/processor/mshv/arm64.rs

```diff
@@ -368,64 +336,77 @@ impl UhProcessor<'_, HypervisorBackedArm64> {
                 UhRunVpError::UnacceptedMemoryAccess(gpa),
             ))
         } else {
-            // TODO SNP: for hardware isolation, if the intercept is due to a guest
+            // TODO: for hardware isolation, if the intercept is due to a guest
             // error, inject a machine check
             self.handle_mmio_exit(dev).await?;
             Ok(())
         }
     }
 }
 
-impl AccessCpuState for UhProcessor<'_, HypervisorBackedArm64> {
+#[derive(Default)]
+pub struct UhCpuStateCache {
+    pc: Option<u64>,
+    sp: Option<u64>,
+    x18: Option<u64>,
+
+    cpsr: Option<Cpsr64>,
+}
+
+impl<T: CpuIo> AccessCpuState for UhEmulationState<'_, '_, T, HypervisorBackedArm64> {
     fn commit(&mut self) {
         let mut expensive_regs = Vec::with_capacity(3);
-        if self.backing.cpu_state.x18_valid {
-            expensive_regs.push((HvArm64RegisterName::X18, self.x(18)));
+        if let Some(x18) = self.cache.x18 {
```

Comment:
Seems unfortunate to have to flush x18 if we only read it and never modified it.